spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/repository/AppRegistrationDao.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.registry.repository;
+
+import org.springframework.cloud.dataflow.core.AppRegistration;
+import org.springframework.cloud.dataflow.core.ApplicationType;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.lang.Nullable;
+
+/**
+ * DAO to access {@link org.springframework.cloud.dataflow.core.AppRegistration}. Contains
+ * predicate specific operations to make filtering based on optional parameters more
+ * efficient.
+ *
+ * @author Siddhant Sorann
+ */
+public interface AppRegistrationDao {
```

Comment:
Instead of creating a new DAO, can we add this method in [here](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/repository/AppRegistrationRepository.java#L33)? or, do you see any issue with that approach?

Reply:
Did you try to address this via JPA repository specific implementation? If so, we would be keen to understand the issues you faced. In case if you would like to take a reference, you can check this repository implementation here in [Skipper code base](https://github.com/spring-cloud/spring-cloud-skipper/blob/805dcf99bfb6fc965fa23712eb814149df8538ea/spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/repository/jpa/ReleaseRepositoryImpl.java#L30) to see a pattern on how you can have repository specific implementation.

Reply:
@ilayaperumalg I decided to use predicates for this to avoid multiple if conditions and JPA repo functions based on all possible combintations of filters being sent via the API. Since in this case we have 4 optional filters that would lead to about 24 such cases.

Reply:
@ilayaperumalg I have made the changes in the implementation according to what you described above, please take a look.

Reply:
@siddhantsorann Thanks for the followup and there was nothing wrong in your earlier changes instead, I was trying to understand the reasonings and more importantly, it would be better to have the same approach in terms of customising the repositories in our codebase. I will review this PR and get back to you.

Reply:
@ilayaperumalg I got it. This approach looks more cleaner and requires lesser changes so it's a better design.